### PR TITLE
TS-4322: Proposal: NetProfileSM

### PR DESCRIPTION
**Pull request by oknet**
This a draft version of NetProfileSM.

The NetProfileSM is attached to a NetVC and called by NetHandler to accomplish Net I/O.

The PR is merged from our 6.0.x branch and it is a hard working.

I will rebase it if the design is accepted.

@SolidWallOfCode @jpeach @shinrich @zwoop @bryancall

### CODE REVIEW

File: iocore/net/Makefile.am [hunk 1]
```diff
@@ -73,13 +76,11 @@ libinknet_a_SOURCES = \
   P_Socks.h \
   P_SSLCertLookup.h \
   P_SSLConfig.h \
-  P_SSLNetAccept.h \
   P_SSLNetProcessor.h \
-  P_SSLNetVConnection.h \
-  P_SSLNextProtocolAccept.h \
-  P_SSLNextProtocolSet.h \
+  I_SSLNextProtocolSet.h \
```
Comment (shinrich) [inline iocore/net/Makefile.am]: Picky but I_SSLNextProtocolSet.h doesn't need to be in list twice.

File: iocore/net/NetProfileSM.cc [hunk 2]
```diff
@@ -1,5 +1,7 @@
 /** @file
 
+  A brief file description
+
```
Comment (shinrich) [inline iocore/net/NetProfileSM.cc]: And the new files should have the brief description filed out.

File: iocore/net/SSLProfileSM.cc [hunk 3]
```diff
@@ -206,20 +275,11 @@ ssl_read_from_net(SSLNetVConnection *sslvc, EThread *lthread, int64_t &ret)
...
-    if (!sslvc->origin_trace) {
-      TraceIn((0 < nread && trace), sslvc->get_remote_addr(), sslvc->get_remote_port(), "WIRE TRACE\tbytes=%d\n%.*s", (int)nread,
-              (int)nread, current_block);
-    } else {
-      char origin_trace_ip[INET6_ADDRSTRLEN];
-      ats_ip_ntop(sslvc->origin_trace_addr, origin_trace_ip, sizeof(origin_trace_ip));
-      TraceIn((0 < nread && trace), sslvc->get_remote_addr(), sslvc->get_remote_port(), "CLIENT %s:%d\ttbytes=%d\n%.*s",
-              origin_trace_ip, sslvc->origin_trace_port, (int)nread, (int)nread, current_block);
-    }
+    Debug("ssl", "[SSLNetProfileSM::read_from_net] nread=%d", (int)nread);
```
Comment (shinrich) [inline iocore/net/SSLProfileSM.cc]: Where did the trace logic go?

**Comment by shinrich** [comment]
At first glance this looks reasonable.  It is a big change, but that logic could use some restructuring.  Perhaps I can get you to sketch out the class hierarchy on a white board during break today.  I also want to pull this branch and work with it a bit.  Then make another pass of comments.

File: iocore/net/I_NetVConnection.h [hunk 4]
```diff
@@ -541,6 +545,31 @@ class NetVConnection : public VConnection
   unsigned int attributes;
   EThread *thread;
 
+  // es - origin_trace associated connections
```
Comment (oknet) [inline iocore/net/I_NetVConnection.h]: yes, pulled from SSLNetVConnection. And put the wiretrace logic into raw_read/write and read/write function that make the code simply and clearly. Just call NetProfileSM::read()/write()/raw_read()/raw_write() and you will get wiretrace feature.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/1024/ for details.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1130/ for details.

File: iocore/net/Makefile.am (hunk 1 above)
Comment (oknet) [inline iocore/net/Makefile.am]: Here, I'm rename P_SSLNextProtocolSet.h to I_SSLNextProtocolSet.h.
It expose the "SSLNextProtocolSet" for "ProtocolProbeSessionAccept" in proxy directory.
The dup one removed.

File: iocore/net/NetProfileSM.cc (hunk 2 above)
Comment (oknet) [inline iocore/net/NetProfileSM.cc]: sorry, I don't know what should be here, just copy the header from UnixNetVConnection.cc.

File: iocore/net/SSLProfileSM.cc (hunk 3 above)
Comment (oknet) [inline iocore/net/SSLProfileSM.cc]: move it into SSLProfileSM::read()/write()

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1131/ for details.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/1025/ for details.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/1026/ for details.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1132/ for details.

**Comment by bryancall** [comment]
Can you please rebase this and we would like to land this for 7.2.0.

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
@oknet What do we want to do with this PR?

**Comment by oknet** [comment]
@bryancall I think that this needs to be broken up into a series of staged commits. because I designed this in 6.0.x branch it is different than master now.

**Comment by bryancall** [comment]
@oknet What do you want to do with this PR?  It has been open for awhile.

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
There hasn't been an update on this PR for over 60 days, so I am closing it for now.  Feel free to reopen the PR when you have updated it.

**Pull request closed without merging**